### DISCUSSION TITLE
Collecting source code rules

### DISCUSSION THREAD

**Discussion by wolfgang-ch** [mytourbook/mytourbook#919] [General]
# Source Code File

Set default settings for source code files

* Text file encoding
* New line delimiter

![image](https://github.com/mytourbook/mytourbook/assets/1283445/8c71c170-6ebd-4da8-b04a-7811cb360cc2)

Existing files can be converted with

![image](https://github.com/mytourbook/mytourbook/assets/1283445/0f0db39f-a93a-4d42-a752-03a699015d26)

# Code Rules

This is a collection of best practices for code rules.

## Transient fields

Transient fields should be located AFTER all fields which are saved in the entity

## Private fields should start with an underscore _

Private fields in inner classes should have a double underscore `__` prefix.

![image](https://user-images.githubusercontent.com/1283445/187604121-75ce7ffa-6683-4c34-8ff3-c4ac6970bddd.png)

**Advantages**
When renaming variables, then you immediately know where it is visible and you can rename it easily or you need Ctrl-Enter to rename it also in other files. The blue var name shows also if it's a local var or a class var.

## No shortcut names 

except when they are used over and over again, e.g.

- `NL` New line
- `UI`  User interface

----

- `gd` GridDataFactory
- `num` Number of
- `pc` Pixel converter
- `sb` StringBuilder
- `tvi` Tree viewer item

or when they are very common, e.g.

- `FPS` Frames per second

## Use lambda expressions and not method references
See https://github.com/mytourbook/mytourbook/discussions/1047

## Text keys from other`Messages.java`

That the externalize strings tool works, only **one** `Messages.java` can be in the import section of a java file. 

Use https://github.com/mytourbook/mytourbook/blob/main/bundles/net.tourbook/src/net/tourbook/OtherMessages.java for messages which are not in `Messages.java` that is bound to the externalize string tool.

## Newline characters \n in `messages.properties`

At the end of a one-line or multi-line text in `messages.properties`, there should be not a newline character added because this is not visible for a translator of this text. When there is a need for a newline at the end of a text, this should be added programmatically. 

There is old code which have not yet applied this rule, see this [discussion](https://github.com/mytourbook/mytourbook/commit/8cb13a399155f970f1e27f859f13e2075d09ed03#r123695728) 

## Source code formatting issues
* https://github.com/mytourbook/mytourbook/discussions/918#discussion-4340787

# Style Rules
* For actions, use "Title Case" https://en.wikipedia.org/wiki/Title_case mainly e.g. 'at', 'of', ... should be lowercase otherwise it looks ugly

# UI

## UI controls

All UI control fields should be sorted as the last fields in a class, when UI controls are available

## <a href="#user-content-common-methods" id="common-methods">Common methods</a>

A UI is created mostly with the same or similar methods

| Method name | Comment
| --- | --- |
| initUI | Setup common listener, ...
| createUI | Create controls
| fillUI | Fill combo boxes, ...
| updateUI | Set values into the controls

**Comment by FJBDev** [comment]
> * Private fields are starting with an underscore _

Unfortunately, the Sonar team [has confirmed it's not possible to enforce that specific rule for private fields only](https://community.sonarsource.com/t/java-field-names-should-comply-with-a-naming-convention-s116/69234/6) but it's OK since you disabled the rule in SonarCloud

**Comment by wolfgang-ch** [comment]
From time to time I update these rules in the top comment, but you cannot receive a notification message when this is done

![image](https://user-images.githubusercontent.com/1283445/189385653-c872b002-9c88-4ef2-b426-2f247e872716.png)

**Reply by FJBDev** [reply]
Good idea because even though I have "subscribed" to this discussion, I don't receive notification of modification

**Comment by FJBDev** [comment]
@wolfgang-ch 
Did you ever test the [CheckStyle ](https://checkstyle.org/eclipse-cs/#!/)Eclipse plugin ? it seems like it could enforce most of the source code rules you wrote here

**Reply by wolfgang-ch** [reply]
Not recently but as I remember, 10+ years ago I tested such a tool and it produced a lot of warnings which was the reason not to use it.

**Reply by FJBDev** [reply]
> Not recently but as I remember, 10+ years ago I tested such a tool and it produced a lot of warnings which was the reason not to use it.

By default, it will produce a lot of warning/errors. But if you [configure ](https://checkstyle.org/eclipse-cs/#!/custom-config)the rules, then it will display only the relevant errors

**Reply by FJBDev** [reply]
I was thinking it would be good to use this CheckStyle plugin because I found it in the Eclipse Marketplace as it is[ one of the most popular extension](https://marketplace.eclipse.org/content/checkstyle-plug)
